x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/FollowIndexAction.java

```diff
@@ -510,13 +526,27 @@ void finalizeResponse() {
 
     static void validate(Request request,
                          IndexMetaData leaderIndex,
-                         IndexMetaData followIndex, MapperService followerMapperService) {
+                         IndexMetaData followIndex,
+                         Map<Integer, String> leaderIndexHistoryUUID,
+                         MapperService followerMapperService) {
         if (leaderIndex == null) {
             throw new IllegalArgumentException("leader index [" + request.leaderIndex + "] does not exist");
         }
         if (followIndex == null) {
             throw new IllegalArgumentException("follow index [" + request.followerIndex + "] does not exist");
         }
+
+        Map<Integer, String> recordedHistoryUUIDs = convert(followIndex.getCustomData(Ccr.CCR_CUSTOM_METADATA_KEY));
+        assert recordedHistoryUUIDs.size() == leaderIndexHistoryUUID.size();
+        for (Map.Entry<Integer, String> entry : leaderIndexHistoryUUID.entrySet()) {
+            String recordedLeaderIndexHistoryUUID = recordedHistoryUUIDs.get(entry.getKey());
+            if (entry.getValue().equals(recordedLeaderIndexHistoryUUID) == false) {
+                throw new IllegalArgumentException("follow index [" + request.followerIndex + "] should reference [" +
```

Comment:
maybe add some info as to how this can happen? (restore from snapshot is a likely cause, I think)